io-sim/src/Control/Monad/IOSim/Types.hs

```diff
@@ -0,0 +1,805 @@
+
+pattern SimTrace :: Time -> ThreadId -> Maybe ThreadLabel -> SimEventType -> SimTrace a
+                 -> SimTrace a
+pattern SimTrace time threadId threadLabel traceEvent trace =
+    Trace.Cons (SimEvent time threadId threadLabel traceEvent)
+               trace
+
+pattern TraceRacesFound :: [ScheduleControl] -> SimTrace a
+                        -> SimTrace a
+pattern TraceRacesFound controls trace =
+    Trace.Cons (SimRacesFound controls)
+               trace
+
+pattern TraceMainReturn :: Time -> a -> [Labelled ThreadId]
+                        -> SimTrace a
+pattern TraceMainReturn time a threads = Trace.Nil (MainReturn time a threads)
+
+pattern TraceMainException :: Time -> SomeException -> [Labelled ThreadId]
+                           -> SimTrace a
+pattern TraceMainException time err threads = Trace.Nil (MainException time err threads)
+
+pattern TraceDeadlock :: Time -> [Labelled ThreadId]
+                      -> SimTrace a
+pattern TraceDeadlock time threads = Trace.Nil (Deadlock time threads)
+
+pattern TraceLoop :: SimTrace a
+pattern TraceLoop = Trace.Nil Loop
+
+{-# COMPLETE SimTrace, TraceMainReturn, TraceMainException, TraceDeadlock, TraceLoop #-}
+{-# COMPLETE Trace,    TraceMainReturn, TraceMainException, TraceDeadlock, TraceLoop #-}
+
+
+data SimEventType
+  = EventSay  String
+  | EventLog  Dynamic
+  | EventMask MaskingState
+
+  | EventThrow          SomeException
+  | EventThrowTo        SomeException ThreadId -- This thread used ThrowTo
+  | EventThrowToBlocked                        -- The ThrowTo blocked
+  | EventThrowToWakeup                         -- The ThrowTo resumed
+  | EventThrowToUnmasked (Labelled ThreadId)   -- A pending ThrowTo was activated
+
+  | EventThreadForked    ThreadId
+  | EventThreadFinished                  -- terminated normally
+  | EventThreadUnhandled SomeException   -- terminated due to unhandled exception
+
+  | EventTxCommitted   [Labelled TVarId] -- tx wrote to these
+                       [TVarId]          -- and created these
+  | EventTxAborted
+  | EventTxBlocked     [Labelled TVarId] -- tx blocked reading these
+  | EventTxWakeup      [Labelled TVarId] -- changed vars causing retry
+
+  | EventTimerCreated   TimeoutId TVarId Time
+  | EventTimerUpdated   TimeoutId        Time
+  | EventTimerCancelled TimeoutId
+  | EventTimerExpired   TimeoutId
+
+  -- the following events are inserted to mark the difference between
+  -- a failed trace and a similar passing trace of the same action
+  | EventThreadSleep                      -- the labelling thread was runnable,
+                                          -- but its execution was delayed
+  | EventThreadWake                       -- until this point
+  deriving Show
+
+type TraceEvent = SimEventType
+{-# DEPRECATED TraceEvent "Use 'SimEventType' instead." #-}
+
+data ThreadId = ThreadId  [Int]
+              | TestThreadId [Int]    -- test threads have higher priority
```

Comment:
Not necessarily. If a test thread spawns another thread WITHOUT calling exploreRaces, then that thread will also be a test thread. You could imagine spawning a mock service, for example, which the SUT interacts with--generally you would not be interested in looking for race conditions in the mock code.

Reply:
I'd prefer a different naming convention then:
```
data ThreadId = ThreadId     [Int] -- the current `TestThreadId`
              | RacyThreadId [Int] -- the current `ThreadId`
```

It seems that `threadRacy :: !Bool` field of `Thread` is redundant.  It would probably be enough to have
```
threadRacy :: Thread -> Bool
threadRace Thread { threadId } = case threadId of ...
```

Reply:
If we agree on this, we'd also need to change how `exploreRaces` is implemented, [ref](https://github.com/input-output-hk/ouroboros-network/pull/3463/files?show-viewed-files=true&file-filters%5B%5D=#diff-3cf14b19cfdfecbb57e97f5261535aa17be65b9b65ef09f10afce417bd857074R572-R574).

Reply:
Thinking about this more, renaming is fine, but getting rid of `threadRacy` flag is not.  Thread ids are used in various data structures and we should change the name once they are set (it could cause wired bugs).

However, I noticed that when we insert threads to `runqueue` we use:
```hs
runqueue' = List.insertBy (comparing Down) tid runqueue
```
a thread with `threadRacy` flag set, might still be a `TestThread` (or after renaming `ThreadId`), and it will be scheduled earlier than expected.